pkg/scrape/discovery/aws/ec2.go

```diff
@@ -0,0 +1,330 @@
+		region, err := metadata.Region()
+		if err != nil {
+			return errors.New("EC2 SD configuration requires a region")
+		}
+		c.Region = region
+	}
+	for _, f := range c.Filters {
+		if len(f.Values) == 0 {
+			return errors.New("EC2 SD configuration filter values cannot be empty")
+		}
+	}
+	return nil
+}
+
+// EC2Discovery periodically performs EC2-SD requests. It implements
+// the Discoverer interface.
+type EC2Discovery struct {
+	*refresh.Discovery
+	logger logrus.FieldLogger
+	cfg    *EC2SDConfig
+	ec2    *ec2.EC2
+
+	// azToAZID maps this account's availability zones to their underlying AZ
+	// ID, e.g. eu-west-2a -> euw2-az2. Refreshes are performed sequentially, so
+	// no locking is required.
+	azToAZID map[string]string
+}
+
+// NewEC2Discovery returns a new EC2Discovery which periodically refreshes its targets.
+func NewEC2Discovery(conf *EC2SDConfig, logger logrus.FieldLogger) *EC2Discovery {
+	d := &EC2Discovery{
+		logger: logger,
+		cfg:    conf,
+	}
+	d.Discovery = refresh.NewDiscovery(
+		logger,
+		"ec2",
+		d.cfg.RefreshInterval,
+		d.refresh,
+	)
+	return d
+}
+
+func (d *EC2Discovery) ec2Client(ctx context.Context) (*ec2.EC2, error) {
+	if d.ec2 != nil {
+		return d.ec2, nil
+	}
+
+	creds := credentials.NewStaticCredentials(d.cfg.AccessKey, d.cfg.SecretKey, "")
+	if d.cfg.AccessKey == "" && d.cfg.SecretKey == "" {
+		creds = nil
+	}
+
+	sess, err := session.NewSessionWithOptions(session.Options{
+		Config: aws.Config{
+			Endpoint:    &d.cfg.Endpoint,
+			Region:      &d.cfg.Region,
+			Credentials: creds,
+		},
+		Profile: d.cfg.Profile,
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "could not create aws session")
+	}
+
+	if d.cfg.RoleARN != "" {
+		creds := stscreds.NewCredentials(sess, d.cfg.RoleARN)
+		d.ec2 = ec2.New(sess, &aws.Config{Credentials: creds})
+	} else {
+		d.ec2 = ec2.New(sess)
+	}
+
+	return d.ec2, nil
+}
+
+func (d *EC2Discovery) refreshAZIDs(ctx context.Context) error {
+	azs, err := d.ec2.DescribeAvailabilityZonesWithContext(ctx, &ec2.DescribeAvailabilityZonesInput{})
+	if err != nil {
+		return err
+	}
+	d.azToAZID = make(map[string]string, len(azs.AvailabilityZones))
+	for _, az := range azs.AvailabilityZones {
+		d.azToAZID[*az.ZoneName] = *az.ZoneId
+	}
+	return nil
+}
+
+func (d *EC2Discovery) refresh(ctx context.Context) ([]*targetgroup.Group, error) {
+	ec2Client, err := d.ec2Client(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	tg := &targetgroup.Group{
+		Source: d.cfg.Region,
+	}
+
+	var filters []*ec2.Filter
+	for _, f := range d.cfg.Filters {
+		filters = append(filters, &ec2.Filter{
+			Name:   aws.String(f.Name),
+			Values: aws.StringSlice(f.Values),
+		})
+	}
+
+	// Only refresh the AZ ID map if we have never been able to build one.
+	// Prometheus requires a reload if AWS adds a new AZ to the region.
+	if d.azToAZID == nil {
+		if err := d.refreshAZIDs(ctx); err != nil {
+			d.logger.WithError(err).Debug("msg", "Unable to describe availability zones")
```

Comment:
Here and elsewhere: logrus has a slightly different API and requires fields to be set explicitly, except for the `message` field:
```suggestion
			d.logger.WithError(err).Debug("Unable to describe availability zones")
```

Reply:
got it, fixed